src/computeScreenCoords.py

```diff
@@ -1,6 +1,70 @@
+from enum import Enum
+import numpy as np
+from sklearn.linear_model import LinearRegression
+import pandas as pd 
+from scipy.interpolate import LinearNDInterpolator
+from typing import List, Tuple
+
+class InterpolationType(Enum):
+    LINEAR = 1
+    RBF_LINEAR = 2
+
+def unpackEyeCoords(eyeCoords: List[List[Tuple]]) -> List[Tuple]:
+    # unpacks [[(1,2),(3,4)], [(5,6), (7,8)]] to [(1,2,3,4), (5,6,7,8)]
+    return [sum(tupList,()) for tupList in eyeCoords]
+
+def unpackScreenCoords(screenCoords):
+    screenXCoords, screenYCoords = zip(*screenCoords)
+    return (list(screenXCoords), list(screenYCoords))
+
+class LinearRegression():
+    def __init__(self, eyeCoords: List[Tuple], screenXCoords: List, screenYCoords: List):
+        df_X = pd.DataFrame(eyeCoords)
+        # TODO: restructure to perform unpacking in models instead of beforehand to account for differences in implementations
+        df_Y = pd.DataFrame(zip(screenXCoords, screenYCoords))
+        self.model = LinearRegression()
+        self.model.fit(df_X, df_Y)
+    def computeScreenCoords(self, eyeCoords):
+        df_X = pd.DataFrame(eyeCoords)
+        prediction = self.model.predict(df_X)
+        return list(prediction.itertuples(index=False, name=None))
+        
+class LinearInterpolator():
+    def __init__(self, eyeCoords: List[Tuple], screenXCoords: List, screenYCoords: List):
+        self.xInterpolator = LinearNDInterpolator(eyeCoords, screenXCoords)
+        self.yInterpolator = LinearNDInterpolator(eyeCoords, screenYCoords)
+    def computeScreenCoords(self, eyeCoords):
+        return (self.xInterpolator(eyeCoords), self.yInterpolator(eyeCoords))
+
+class RBFInterpolator():
+    # def __init__(self, eyeCoords: List[Tuple], screenXCoords: List, screenYCoords: List, kernel=None, smoothing=None):
+    #     self.kernel = kernel
+    #     self.smoothing = smoothing
+    #     self.rbf = RBFInterpolator()
     pass
 
+class Interpolator():
+    def __init(self):
+        self.interpolator = None
+
+    def calibrateInterpolator(self, eyeCoords: List[List[Tuple]], screenCoords: List[Tuple], interpType = InterpolationType.LINEAR):
+        # eyeCoords of shape [[(x1.1,y1.1),(x1.2,y1.2)], [(x2.1,y2.1, x2.2, y2.2)], etc.]
+        # screenCoords of shape [(x1,y1), (x2,y2), (x3,y3), etc]
+        if interpType == InterpolationType.LINEAR:
+            unpackedEyeCoords = unpackEyeCoords(eyeCoords)
+            unpackedXScreenCoords, unpackedYScreenCoords = unpackScreenCoords(screenCoords)
+            self.interpolator = LinearInterpolator(unpackedEyeCoords, unpackedXScreenCoords, unpackedYScreenCoords)
+        return
+    
+    def computeScreenCoords(self, eyeCoords) -> Tuple:
+        # eyeCoords of shape [x1,y1,x2,y2]
+        unpackedEyeCoords = unpackEyeCoords(eyeCoords)
+        if self.interpolator is None:
+            raise ValueError('Interpolator not calibrated yet.') # TODO: create exception class to be more accurate? not sure
```

Comment:
Depending on how you implement the entry point for your code (see comment below), you may want the `if self.interpolator is None` guard to call calibrateInterpolator() so that computeScreenCoords() can be called from main regardless of whether it's been calibrated already (i.e. calibrate the first time, then every call after that would skip calibration).

Reply:
Good point, but that would require passing in the parameters for calibration to the computation. Ideally the interpolator should already be calibrated. It's currently set up like this since I wasn't sure if we would have the calibration data ready when the interpolator is initially instantiated (and therefore it would be a separate call), but maybe I can add the calibration call to the initialization. Thoughts?

Reply:
To my eyes, it looks like I would end up calling `calibrateInterpolator` from my code once the user has completed calibration with the circles. Is that right, or am I reading this the wrong way?

Reply:
That's a good point, we should probabally wait until Zach finishes the calibration part to see how best to integrate it. Potentially, we could read the data from a file here and perform the calibration if it hasn't already been set, but we would also need a way to disregard past calibration data if the user decides to re-calibrate, so maybe it would be best to initiate calibration in some sort of a UI callback after the user finishes. But I do think having the same interface whether or not the models have been trained is a good idea, so maybe the if guard will end up being "if calibration file doesn't exist, then throw exception" or something like that, instead of "if the model hasn't been calibrated yet, then throw exception"